File: COURS/MLB/LABS/CancerRelapse_EnsembleMethods_2.R
# ===================================================
#      Prediction of a categorical variable
# ===================================================
library(pROC)


# Cancer Relapse dataset
X.all = read.table("https://perso.univ-rennes1.fr/valerie.monbet/MachineLearning/Ex2_breast_cancer_ER_train_set_for_other_methods.gct",skip=2,header=TRUE,sep = "\t")
label = read.table("https://perso.univ-rennes1.fr/valerie.monbet/MachineLearning/Ex2_breast_cancer_ER_train_set_class_labels.cls",skip=2,header=FALSE,sep = " ")


#chemin = "~/Documents/ENSEIGNEMENT/MACHINE LEARNING/Master BMC/PROJETS/Ex_datasets/Ex2/"
#X.all = read.table(paste(chemin,"Ex2_breast_cancer_ER_train_set_for_other_methods.gct",sep=""),skip=2,header=TRUE,sep = "\t")
#label = read.table(paste(chemin,"Ex2_breast_cancer_ER_train_set_class_labels.cls",sep=""),skip=2,header=FALSE,sep = " ")

X.desc = X.all[,1]
X = t(X.all[,-c(1,2)])
colnames(X) = X.desc

y = factor(label,levels=c(0,1))


train_test_split <- function (X,y,test_size=.25,random_state=NULL){
  # Extraction of a train and a test datasets.
  #
  # Inputs : X, y : data
  #          test_size : a fraction (<1) of the total set or a number of samples (integer) ; default 0.25
  #          random_state : if equal to an interger, it fixes the random seed ; defaut NULL
  # Outputs : X_train, X_test, y_train, y_test
  #
  n = nrow(X)
  if (test_size>1){test_size=test_size/n}
  if (!is.null(random_state)){set.seed(random_state)}
  itest=sample(1:n,round(n*test_size))
  itrain=setdiff(1:n,itest)
  Xtrain=X[itrain,]
  Xtest=X[itest,]
  ytrain=y[itrain]
  ytest=y[itest]
  return(list(X_train=Xtrain,X_test=Xtest,y_train=ytrain,y_test=ytest))
}


# =======================================================
# Function to find the genes with the highest SNR
genes_SNR = function(X,y){
  # - Inputs
  # X = dataset
  # y = labels (a boolean or factor variable)
  # - Output
  # SNR
  labels = unique(y)
  n = dim(X)[1]
  p = dim(X)[2]
  K = length(labels)
  
  means = matrix(0,length(labels),dim(X)[2])
  std = apply(X,2,sd)
  
  for (k in 1:K){means[k,] = apply(X[y==labels[k],],2,mean)}
  SNR = apply(matrix(abs(apply(means,2,diff)),K-1,p),2,max)/std
  return(SNR)
}

# =======================================================
# adaboost =====================================================================
library(adabag) # boosting

B = 30 # number of samples for the cross validation

err_adab = NULL
prob_adab = NULL
ytest_adab = NULL
for (b in 1:B){
  print(paste("b =",b))
  # sampling for train, test sets
  samples = train_test_split(X,y,test_size=0.25)
  X_train=samples$X_train
  X_test=samples$X_test
  y_train=samples$y_train
  y_test=samples$y_test
  ytest_adab = c(ytest_adab,y_test)
  # standardization
  mk=apply(X_train,2,mean)
  sk=apply(X_train,2,sd)
  X_train = scale(X_train,center=mk,scale=sk)
  X_test = scale(X_test,center=mk,scale=sk)
  data_train = data.frame(y_train,X_train)
  data_test = data.frame(y_test,X_test)
  colnames(data_train)[1] = "y"
  colnames(data_test)=colnames(data_train)
  
  # learning decision tree mode
  adaboost = boosting(y~.,data=data_train,mfinal=20,control=rpart.control(maxdepth=5, minsplit=5))
  y_pred= predict(adaboost,data_test,type="class") 
  err_adab[b] = mean((y_test!=y_pred)) 
  prob_adab = c(prob_adab,predict(adaboost,data_test)$prob[,2])
}  
err_adab_all = mean(err_adab) # global error
print(paste("Mean classification adaboost, tree = ",round(err_adab_all,2)))

res_roc = roc(ytest_adab,prob_adab)
plot(res_roc) # ROC curve
text(.5,.2,paste("AUC =",round(res_roc$auc,2)))


# Variable importance
plot(sort(adaboost$importance))
print("The most importante variables for prediction are")
X.desc[which(adaboost$importance>3)]

# Gradient Boosting =====================================================================
library(gbm) # bug???

# One first example
# sampling for train, test sets
samples = train_test_split(X,y,test_size=0.25)
X_train=samples$X_train
X_test=samples$X_test
y_train=samples$y_train
y_test=samples$y_test
ytest_adab = c(ytest_adab,y_test)
# standardization
mk=apply(X_train,2,mean)
sk=apply(X_train,2,sd)
X_train = scale(X_train,center=mk,scale=sk)
X_test = scale(X_test,center=mk,scale=sk)
data_train = data.frame(y_train,X_train)
data_test = data.frame(y_test,X_test)
colnames(data_train)[1] = "y"
colnames(data_test)=colnames(data_train)
data_train$y = as.numeric(data_train$y)
data_train$y = data_train$y-1
# learning decision tree mode
gb = gbm(y~.,data=data_train, shrinkage=0.01, distribution = 'bernoulli', cv.folds=5, n.trees=500, verbose=F)
best.iter = gbm.perf(gb,method="cv")

B = 30 # number of samples for the cross validation

err_gbm = NULL
prob_gbm = NULL
ytest_gbm = NULL
for (b in 1:B){
  print(paste("b =",b))
  # sampling for train, test sets
  samples = train_test_split(X,y,test_size=0.25)
  X_train=samples$X_train
  X_test=samples$X_test
  y_train=samples$y_train
  y_test=samples$y_test
  ytest_gbm = c(ytest_gbm,y_test)
  # standardization
  mk=apply(X_train,2,mean)
  sk=apply(X_train,2,sd)
  X_train = scale(X_train,center=mk,scale=sk)
  X_test = scale(X_test,center=mk,scale=sk)
  data_train = data.frame(y_train,X_train)
  data_test = data.frame(y_test,X_test)
  colnames(data_train)[1] = "y"
  colnames(data_test)=colnames(data_train)
  data_train$y = as.numeric(data_train$y)
  data_train$y = data_train$y-1
  # learning decision tree mode
  gb = gbm(y~.,data=data_train, shrinkage=0.01, distribution = 'bernoulli', n.trees=500, verbose=F)
  prob_gbm = c(prob_gbm,predict(gb,data_test,n.trees=500,type="response"))
}  

res_roc = roc(ytest_gbm,prob_gbm)
plot(res_roc) # ROC curve
text(.5,.2,paste("AUC =",round(res_roc$auc,2)))



# eXtreme Gradient Boosting =====================================================================
library(xgboost)

ygb = as.numeric(y)-1

B = 30 # number of samples for the cross validation

err_xgb = NULL
prob_xgb = NULL
ytest_xgb = NULL
for (b in 1:B){
  # sampling for train, test sets
  samples = train_test_split(X,ygb,test_size=0.25)
  X_train=samples$X_train
  X_test=samples$X_test
  y_train=samples$y_train
  y_test=samples$y_test
  ytest_xgb = c(ytest_xgb,y_test)
  # standardization
  # standardization
  mk=apply(X_train,2,mean)
  sk=apply(X_train,2,sd)
  X_train = scale(X_train,center=mk,scale=sk)
  X_test = scale(X_test,center=mk,scale=sk)
  dtrain <- xgb.DMatrix(as.matrix(X_train), label = y_train)
  dtest <- xgb.DMatrix(as.matrix(X_test), label = y_test)
  
  param <- list(max_depth = 2, eta = 1, silent = 1, nthread = 2,alpha=.5,
                objective = "reg:logistic")
  bst <- xgb.train(param, dtrain, nrounds = 30)
  # maxdepth = maximal depth of each individual tree
  # minsplit = minimum number of observations in a node to allow a split
  y_pred= predict(bst,dtest) 
  prob_xgb = c(prob_xgb,y_pred)
  
}

res_roc = roc(ytest_xgb,prob_xgb)
plot(res_roc) # ROC curve
text(.5,.2,paste("AUC =",round(res_roc$auc,2)))

File: COURS/MLB/LABS/CancerRelapse_EnsembleMethods_2.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Sun Jul 29 13:13:23 2018

@author: valerie
"""

# ===================================================
#      Prediction of a categorical variable
# ===================================================
import numpy as np
import matplotlib.pyplot as plt
from urllib.request import urlopen

# Cancer Relaps dataset
url = "https://perso.univ-rennes1.fr/valerie.monbet/MachineLearning/Ex2_breast_cancer_ER_train_set_for_other_methods.gct"
X = np.loadtxt(urlopen(url),skiprows=3,usecols=range(2,99))
X = X.T
names = np.genfromtxt(urlopen(url),dtype='str',usecols=1,skip_header=3)

url = "https://perso.univ-rennes1.fr/valerie.monbet/MachineLearning/Ex2_breast_cancer_ER_train_set_class_labels.cls"
label = np.loadtxt(urlopen(url),skiprows=2)

y = label

## ============================================================================
## Import utilities
from sklearn.model_selection import train_test_split
from sklearn.metrics import roc_curve, auc

def ROC(y_test,y_score,methodName=" ",plot=True):

    ntest = np.size(y_test,0)
    B = np.size(y_test,1)
    fpr, tpr, _ = roc_curve(np.reshape(y_test,B*ntest), np.reshape(y_score,B*ntest))
#    if len(fpr)<3:
#        print("Problem: len(fpr) is lower than 3")
#        return
    roc_auc = auc(fpr, tpr)

    if plot:
        lw = 2
        plt.plot(fpr, tpr, color='darkorange',
            lw=lw, label='ROC curve (area = %0.2f)' % roc_auc)
        plt.plot([0, 1], [0, 1], color='navy', lw=lw, linestyle='--')
        plt.xlim([0.0, 1.0])
        plt.ylim([0.0, 1.05])
        plt.xlabel('False Positive Rate')
        plt.ylabel('True Positive Rate')
        plt.title(methodName)
        plt.legend(loc="lower right")
        plt.show()
    return(roc_auc)

# For genes swith best SNR selection ===========================================

def genes_SNR(X,y):
  # - Inputs
  # X = dataset
  # y = labels (a boolean or factor variable)
  # - Output
  # SNR
  labels = np.unique(y)
  n,p = X.shape
  K = len(labels)
  
  means = np.zeros((K,p))
  sd = np.std(X,axis=0)
  
  for k in range(K):
      means[k,] = np.mean(X[y==labels[k],:],axis=0)
  SNR = np.max(np.reshape(np.abs(np.diff(means,axis=0)),(K-1,p)),axis=0)/sd
  return(SNR)

# Adaboost ============================================================
 
from sklearn.ensemble import AdaBoostClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split


B = 30
n_test = 9
y_score = np.zeros([n_test,B]) 
y_test_all = np.zeros([n_test,B]) 
for b in range(B): 
    X_train, X_test, y_train, y_test = train_test_split(X,label,test_size=n_test)
    mk=np.mean(X_train,axis=0)
    sk=np.maximum(np.std(X_train,axis=0),10*np.finfo(float).eps)
    X_train, X_test = np.add(X_train,-mk), np.add(X_test,-mk)
    X_train, X_test = np.multiply(X_train,1/sk),np.multiply(X_test,1/sk)
    clf = AdaBoostClassifier(DecisionTreeClassifier(max_depth=2),n_estimators=20)
    clf.fit(X_train,y_train)
    y_score[:,b] = clf.predict_proba(X_test)[:,1]
    y_test_all[:,b] = y_test
              
ROC(y_test_all,y_score,"AdaBoost")

importances = clf.feature_importances_
indices = np.argsort(importances)[::-1]


print("Feature ranking (20 first:")

for f in range(20):
    print("%d. feature %d (%f)" % (f + 1, indices[f], importances[indices[f]]))

# Plot the feature importances of the last forest
plt.figure(1)
plt.clf()
plt.title("Feature importances (20 first)")
plt.barh(range(20), importances[indices][:20],
       color="r", align="center")
plt.yticks(range(20), [names[i] for i in indices[:20]])
plt.show()


# Gradient Boosting ============================================================
from sklearn.ensemble import GradientBoostingClassifier

B = 30
n_test = 9
y_score = np.zeros([n_test,B]) 
y_test_all = np.zeros([n_test,B]) 
for b in range(B): 
    X_train, X_test, y_train, y_test = train_test_split(X,label,test_size=n_test)
    mk=np.mean(X_train,axis=0)
    sk=np.maximum(np.std(X_train,axis=0),10*np.finfo(float).eps)
    X_train, X_test = np.add(X_train,-mk), np.add(X_test,-mk)
    X_train, X_test = np.multiply(X_train,1/sk),np.multiply(X_test,1/sk)
    clf = GradientBoostingClassifier(loss='exponential',n_estimators=200,max_features=5)
    clf.fit(X_train,y_train)
    y_score[:,b] = clf.predict_proba(X_test)[:,1]
    y_test_all[:,b] = y_test
              
ROC(y_test_all,y_score,"Gradient Boosting")

importances = clf.feature_importances_
indices = np.argsort(importances)[::-1]


print("Feature ranking (20 first:")

for f in range(20):
    print("%d. feature %d (%f)" % (f + 1, indices[f], importances[indices[f]]))

# Plot the feature importances of the last forest
plt.figure(1)
plt.clf()
plt.title("Feature importances (20 first)")
plt.barh(range(20), importances[indices][:20],
       color="r", align="center")
plt.yticks(range(20), [names[i] for i in indices[:20]])
plt.show()


# eXtreme Gradient Boosting ============================================================
import xgboost as xgb

B = 30
n_test = 9
y_pred_proba = []
ytest_xgb = [] 
ncol = X.shape[1]
for b in range(B): 
    X_train, X_test, y_train, y_test = train_test_split(X,label,test_size=n_test)
    ytest_xgb.append(y_test)
    mk=np.mean(X_train,axis=0)
    sk=np.maximum(np.std(X_train,axis=0),10*np.finfo(float).eps)
    X_train, X_test = np.add(X_train,-mk), np.add(X_test,-mk)
    X_train, X_test = np.multiply(X_train,1/sk),np.multiply(X_test,1/sk)
    dtrain = xgb.DMatrix(data=X_train,label=y_train)
    dtest = xgb.DMatrix(data=X_test,label=y_test)
    params = {'n_estimators': 100, 'max_depth': 2, 'min_samples_split': 5,
               'eta': 1,'objective':'reg:logistic','alpha': .1,'silent': 1,
               'early_stopping_rounds': 3}
    watchlist = [(dtest, 'eval')]
    num_round=50
    bst = xgb.train(params,dtrain,num_round,watchlist)
    y_pred_proba.append(bst.predict(dtest))

              
ROC(ytest_xgb,y_pred_proba,"XGB")

##
print('XGBoost with grid search')
# play with these params
params={
    'max_depth': [2,5,9], #[3,4,5,6,7,8,9], # 5 is good but takes too long in kaggle env
    'subsample': [.05,.1,.5], 
    'colsample_bytree': [.5,.7], 
    'n_estimators': [10,100,1000], #[1000,2000,3000]
    'reg_alpha': [0.03,.1,.5,1] #[0.01, 0.02, 0.03, 0.04]
}

xgb_clf = xgb.XGBClassifier()
clf = GridSearchCV(xgb_clf,
                  params,
                  cv=ShuffleSplit(test_size=.25,
                                   n_splits=30, random_state=1), 
                  scoring = "accuracy", 
                  refit=False,n_jobs=-1)
clf.fit(X,y)
print("Best parameters :",rs.best_params_)
print("Best accuracy :",np.round(np.sqrt(-rs.best_score_),2))

